tools/mpy-tool.py

```diff
@@ -152,21 +152,58 @@ def decode_uint(bytecode, ip):
             break
     return ip, unum
 
+def read_compressed_prelude(read_byte):
+    z = read_byte()
+    # xSSSSEAA
+    n_state = (z >> 3) & 0xf;                                   \
+    n_exc_stack = (z >> 2) & 0x1;                               \
+    scope_flags = 0;                                            \
+    n_pos_args = z & 0x3;                                       \
+    n_kwonly_args = 0;                                          \
+    n_def_pos_args = 0;                                         \
+    n = 0
+    while z & 0x80:
+        z = read_byte()
+        # xFSSKAED
+        n_state |= (z & 0x30) << (2 * n);                       \
+        n_exc_stack |= (z & 0x02) << n;                         \
+        scope_flags |= ((z & 0x40) >> 6) << n;                  \
+        n_pos_args |= (z & 0x4) << n;                           \
+        n_kwonly_args |= ((z & 0x08) >> 3) << n;                \
+        n_def_pos_args |= (z & 0x1) << n;                       \
+        n += 1
+    n_state += 0*n_pos_args + 1;                                  \
+
+    return n_state, n_exc_stack, scope_flags, n_pos_args, n_kwonly_args, n_def_pos_args
+
+def read_compressed_prelude2(read_byte):
+    I = 0
+    C = 0
+    n = 0
+    while True:
+        z = read_byte()
+        # xIIIIIIC
+        I |= ((z & 0x7e) >> 1) << (6 * n)
+        C |= (z & 1) << n
+        if not (z & 0x80):
+            break
+        n += 1
+    return I, C
+
 def extract_prelude(bytecode, ip):
-    ip, n_state = decode_uint(bytecode, ip)
-    ip, n_exc_stack = decode_uint(bytecode, ip)
-    scope_flags = bytecode[ip]; ip += 1
-    n_pos_args = bytecode[ip]; ip += 1
-    n_kwonly_args = bytecode[ip]; ip += 1
-    n_def_pos_args = bytecode[ip]; ip += 1
-    ip2, code_info_size = decode_uint(bytecode, ip)
-    ip += code_info_size
-    while bytecode[ip] != 0xff:
+    def local_read_byte():
+        nonlocal bytecode, ip
```

Comment:
I don't know if we're intended to keep `tools/mpy-tool.py` Python 2 compatible, but `nonlocal` breaks that if so.

Reply:
Yeah I think it's good to retain Python 2 compatibility for a while.  Thanks for pointing this out.